### ISSUE TITLE
Joystick bug(s) on linux

### ISSUE THREAD

**Issue by punto-** [godotengine/godot#3727]
(ping @Hinsbart) I'm seeing this behavior (with a PS3 pad on linux)
- press a button, which triggers a call to load() that blocks the main thread for 1 second
- during that second, release the button
- after the load() is finished and the main loop is back to number, the release event doesn't seem to be generated
- press the button again, the event is not generated again
- release the button, the release event is generated

There seems to be 2 bugs, the first lost event is during the time where the main thread is blocked (for example while loading a new level, etc). The 2nd lost event, could it be because the engine thinks the button is already pressed so it omits it? This is ok to do for axis values and mouse movement, but we shouldn't do it for buttons.. We want to send every event that happened, even if the end result is the same state that we previously had. Users are not fast enough, so we won't have more than 2 or 3 events from the same button in one frame (and in that case, the FPS has to be _very_ low).

**Comment by Hinsbart** [comment]
Hmm, could you upload an example for this maybe? I can't reproduce it. For me, the release event is being sent after load() finished. Seems to work for OS.delay_msec() too.

Code was the following:
```
func _input(event):
	if event.type == InputEvent.JOYSTICK_BUTTON and event.button_index == JOY_XBOX_A:
		if event.is_pressed():
			print("A pressed")
			#OS.delay_msec(1000)
			var large_sound = load("test.wav")
			print("block finished")
		else:
			print("A released")
```

**Comment by punto-** [comment]
I can reproduce with that code using delay_msec().. Could it be a problem
specifically with the PS3 pad? Remember you have to release the button
during the delay_msec, while the thread is blocked..

In any case, the second event I'm missing is definitely getting caught by
the check in input_default.cpp:579 that compares it with the current state
(the first one never even gets there).. Is there a particular reason for
that check? The OS should be taking care of that already, if it sends an
even that doesn't match our state there must be a reason.. I'm fine with
sending it to the game..

**Comment by vnen** [comment]
I tested this with my X360 controller on Linux and while I do receive a released event after the block is finished, it ignores anything I did while the thread was blocked. So if I release the button while the thread is blocked and then press it again, it behaves like nothing happened.

**Comment by punto-** [comment]
but if you only release while the thread is blocked you get the event later?

**Comment by Hinsbart** [comment]
@punto- I tested it with more pads and finally found one which does this. Still no idea why tho.

About  input_default.cpp:579, not every platform has event-based joystick input. On xinput/dinput/html5 we only get a struct filled with the current state that is sent to InputDefault each frame.
So this check decides whether to actually send an event.
Of course each platform could perform this themselves, but I'd rather have it in one place.
The reason this triggers in the first place, is that the first event is lost.

Would reading events from the joystick thread instead of the main one help?

**Comment by vnen** [comment]
> but if you only release while the thread is blocked you get the event later?

Yes, I do get the release event later.

**Comment by Hinsbart** [comment]
Can you try [this](https://github.com/Hinsbart/godot/tree/joy_thread) branch?

**Comment by punto-** [comment]
Downloading the joy_thread branch.

If some platforms only provide a state, won't we have the same problem with
losing events? If a user presses and released a button too fast (and the
FPS are low enough) we'll lose those events. This can a problem for certain
games (like fighting games, or games party games where the fastest button
masher wins). Maybe a thread (with something like a queue limit) is
necessary..

 Also are you sure they _only_ provide a state? Most APIs I've seen provide
the state as a helper, but they also have a way to retrieve events..

**Comment by punto-** [comment]
Same behavior with the joy_thread branch, I think I have the right code (my
joystick_linux.cpp is the same as the one on that repository at least..) It
looks like the reading thread is synched with the main thread tho, I put a
print in read_events() and it stops printing when the main thread is
blocked..

**Comment by Hinsbart** [comment]
Ok yeah. I also thought about this shortly after posting that.
I'm not that familiar with multithreading. So what would be the best way to delay processing in the thread (to the framerate?) without synching with the main thread?

**Comment by punto-** [comment]
You could sleep for 16ms, use a ringbuffer or similar to send events to the
main thread..

Do we know why the event is being lost tho? I played around with some test
code, and I noticed that if I sleep for 320ms when the button is pressed,
the event is received.. Is that number important? The events have to be
waiting on some queue or something..

**Comment by akien-mga** [comment]
What's the status on this issue guys? :)

**Comment by akien-mga** [comment]
Ping :)

**Comment by HummusSamurai** [comment]
Can anyone confirm this still happens?

**Comment by HummusSamurai** [comment]
I expect this was corrected by the rewrite of InputMap.

**Comment by akien-mga** [comment]
At any rate I never got any update nor confirmation that there is still stuff to do, so let's close it. If there are still joystick issues on Linux, it would be best to test in 3.0 and open new, better focused issues.

**Comment by Hinsbart** [comment]
Nothing changed in the platform code that handles this, so I'm positive that this issue is still valid.

> I suspect this was corrected by the rewrite of InputMap and should be obsolete.

InputMap is not related to this ^^

> At any rate I never got any update nor confirmation that there is still stuff to do

We need to move the joypad handling to a thread so we won't miss events.
That said, I'm fine with opening a new issue about this.

**Comment by akien-mga** [comment]
> We need to move the joypad handling to a thread so we won't miss events.
That said, I'm fine with opening a new issue about this.

Well, let's reopen then if it's confirmed valid.

**Comment by reduz** [comment]
I would close this, as nothing really happened in years.. I guess also no one is using PS3 pads..

**Comment by akien-mga** [comment]
Agreed. If someone experiences this issue/wants to work on fixing it, we can reopen.
